File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        print(f'{dp}\n')

        for i in range(n - 2, -1, -1):
            print(f'i: {i}')
            for j in range(i + 1, n):
                if nums[i] < nums[j]:
                    dp[i] = max(dp[i], 1 + dp[j])
                    print(f'j: {j}, dp: {dp}')

            print()

        return max(dp)

File: Data Structures & Algorithms/sliding-window-maximum/submission-3.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int> res(n - k + 1);
        deque<int> q;
        int l = 0;

        for (int r = 0; r < n; r++) {
            while (!q.empty() && nums[q.back()] < nums[r]) {
                q.pop_back();
            }
            q.push_back(r);

            if (l > q.front()) {
                q.pop_front();
            }

            if (r + 1 >= k) {
                res[l] = nums[q.front()];
                l++;
            }
        }

        return res;
    }
};
